File: Reference/tests/Brontide.Reference.Studio.Tests/ComponentProviderTrustCadenceRecoveryTests.cs
using System.Text.Json;

namespace Brontide.Reference.Studio.Tests;

public sealed class ComponentProviderTrustCadenceRecoveryTests
{
    private sealed class FakeCycle(Func<DateTimeOffset, ProviderServingTrustCycleResult> run)
        : IProviderServingTrustCycle
    {
        public int Calls { get; private set; }

        public Task<ProviderServingTrustCycleResult> RunAsync(
            DateTimeOffset now,
            CancellationToken cancellationToken)
        {
            Calls++;
            return Task.FromResult(run(now));
        }
    }

    private sealed class FakeDelay : IProviderServingTrustCadenceDelay
    {
        public int Calls { get; private set; }

        public Task<DateTimeOffset> DelayAsync(
            DateTimeOffset now,
            TimeSpan duration,
            CancellationToken cancellationToken)
        {
            Calls++;
            return Task.FromResult(now + duration);
        }
    }

    private sealed class TemporaryJournal : IDisposable
    {
        private readonly string root = System.IO.Path.Combine(
            System.IO.Path.GetTempPath(), $"brontide-cbi48-{Guid.NewGuid():N}");
        public string Path => System.IO.Path.Combine(root, "cadence.bin");
        public void Dispose()
        {
            try { if (Directory.Exists(root)) Directory.Delete(root, true); }
            catch (IOException) { }
            catch (UnauthorizedAccessException) { }
        }
    }

    private static readonly ProviderTrustCadenceRunId RunId =
        ProviderTrustCadenceRunId.Create("cadence-run.test.1");

    private static readonly ProviderServingTrustCadenceSchedule Schedule =
        ProviderServingTrustCadenceSchedule.Create(2, TimeSpan.FromSeconds(5));

    private static readonly DateTimeOffset Start =
        new(2026, 8, 5, 8, 0, 0, TimeSpan.Zero);

    private static ProviderServingTrustCycleResult CycleResult(string code) =>
        new(code, null!, null, 0);

    private static DurableProviderTrustCadenceJournal Establish(string path) =>
        DurableProviderTrustCadenceJournal.Establish(path, RunId, Schedule, Start).Journal!;

    [Test]
    public void Cbi48_C1_a_durable_run_is_bounded_and_distinctly_identified()
    {
        using var temporary = new TemporaryJournal();
        var established = DurableProviderTrustCadenceJournal.Establish(
            temporary.Path, RunId, Schedule, Start);
        var duplicate = DurableProviderTrustCadenceJournal.Establish(
            temporary.Path, RunId, Schedule, Start);
        var mismatch = DurableProviderTrustCadenceJournal.Open(
            temporary.Path, ProviderTrustCadenceRunId.Create("cadence-run.other"));

        Assert.Multiple(() =>
        {
            Assert.That(established.Code, Is.EqualTo("durable-cadence-established"));
            Assert.That(established.Journal!.Snapshot.RunIdentity, Is.EqualTo(RunId));
            Assert.That(duplicate.Code, Is.EqualTo("durable-cadence-already-exists"));
            Assert.That(mismatch.Code, Is.EqualTo("durable-cadence-run-mismatch"));
            Assert.Throws<ArgumentException>(() => DurableProviderTrustCadenceJournal.Open(
                temporary.Path, default));
        });
    }

    [Test]
    public void Cbi48_C2_every_transition_is_atomic_and_integrity_checked()
    {
        using var temporary = new TemporaryJournal();
        var journal = Establish(temporary.Path);
        Assert.That(journal.BeginCycle().Code, Is.EqualTo("durable-cadence-cycle-started"));
        var bytes = File.ReadAllBytes(temporary.Path);
        bytes[^1] ^= 0xff;
        File.WriteAllBytes(temporary.Path, bytes);

        Assert.That(DurableProviderTrustCadenceJournal.Open(temporary.Path, RunId).Code,
            Is.EqualTo("durable-cadence-corrupt"));
    }

    [Test]
    public async Task Cbi48_C3_in_flight_state_precedes_the_effectful_cycle()
    {
        using var temporary = new TemporaryJournal();
        var journal = Establish(temporary.Path);
        var observed = "";
        var cycle = new FakeCycle(_ =>
        {
            observed = DurableProviderTrustCadenceJournal.Open(temporary.Path, RunId).Code;
            return CycleResult("provider-trust-cycle-stopped");
        });

        await ProviderTrustCadenceRecovery.AdvanceAsync(
            journal, cycle, new FakeDelay(), CancellationToken.None);

        Assert.Multiple(() =>
        {
            Assert.That(observed, Is.EqualTo("durable-cadence-indeterminate"));
            Assert.That(cycle.Calls, Is.EqualTo(1));
            Assert.That(journal.Snapshot.Code, Is.EqualTo("durable-cadence-stopped"));
        });

        using var failedWriteTemporary = new TemporaryJournal();
        var failedWrite = Establish(failedWriteTemporary.Path);
        File.Delete(failedWriteTemporary.Path);
        Directory.CreateDirectory(failedWriteTemporary.Path);
        var forbiddenCycle = new FakeCycle(_ => CycleResult("provider-trust-cycle-current"));
        var refused = await ProviderTrustCadenceRecovery.AdvanceAsync(
            failedWrite, forbiddenCycle, new FakeDelay(), CancellationToken.None);
        Assert.Multiple(() =>
        {
            Assert.That(refused.Code, Is.EqualTo("durable-cadence-write-failed"));
            Assert.That(forbiddenCycle.Calls, Is.Zero);
        });
    }

    [Test]
    public async Task Cbi48_C4_completed_work_resumes_from_the_next_clean_boundary()
    {
        using var temporary = new TemporaryJournal();
        var first = Establish(temporary.Path);
        var firstCycle = new FakeCycle(_ => CycleResult("provider-trust-cycle-current"));
        await ProviderTrustCadenceRecovery.AdvanceAsync(
            first, firstCycle, new FakeDelay(), CancellationToken.None);

        var waitingBytes = File.ReadAllBytes(temporary.Path);
        using (var canceled = new CancellationTokenSource())
        {
            canceled.Cancel();
            var canceledCycle = new FakeCycle(_ => CycleResult("provider-trust-cycle-current"));
            var canceledDelay = new FakeDelay();
            var canceledResult = await ProviderTrustCadenceRecovery.AdvanceAsync(
                first, canceledCycle, canceledDelay, canceled.Token);
            Assert.Multiple(() =>
            {
                Assert.That(canceledResult.Code, Is.EqualTo("durable-cadence-wait-canceled"));
                Assert.That(canceledCycle.Calls, Is.Zero);
                Assert.That(canceledDelay.Calls, Is.Zero);
                Assert.That(File.ReadAllBytes(temporary.Path), Is.EqualTo(waitingBytes));
            });
        }

        var recovered = DurableProviderTrustCadenceJournal.Open(temporary.Path, RunId).Journal!;
        var secondCycle = new FakeCycle(_ => CycleResult("provider-trust-cycle-current"));
        var delay = new FakeDelay();
        await ProviderTrustCadenceRecovery.AdvanceAsync(
            recovered, secondCycle, delay, CancellationToken.None);

        Assert.Multiple(() =>
        {
            Assert.That(firstCycle.Calls, Is.EqualTo(1));
            Assert.That(secondCycle.Calls, Is.EqualTo(1));
            Assert.That(delay.Calls, Is.EqualTo(1));
            Assert.That(recovered.Snapshot.Code, Is.EqualTo("durable-cadence-complete"));
            Assert.That(recovered.Snapshot.Cycles.Select(item => item.Index), Is.EqualTo(new[] { 0, 1 }));
        });
    }

    [Test]
    public async Task Cbi48_C5_an_interrupted_effect_is_indeterminate_and_inert()
    {
        using var temporary = new TemporaryJournal();
        var journal = Establish(temporary.Path);
        journal.BeginCycle();
        var before = File.ReadAllBytes(temporary.Path);
        var recovered = DurableProviderTrustCadenceJournal.Open(temporary.Path, RunId);
        var cycle = new FakeCycle(_ => CycleResult("provider-trust-cycle-current"));
        var delay = new FakeDelay();

        var result = await ProviderTrustCadenceRecovery.AdvanceAsync(
            recovered.Journal!, cycle, delay, CancellationToken.None);

        Assert.Multiple(() =>
        {
            Assert.That(recovered.Code, Is.EqualTo("durable-cadence-indeterminate"));
            Assert.That(result.Code, Is.EqualTo("durable-cadence-indeterminate"));
            Assert.That(cycle.Calls, Is.Zero);
            Assert.That(delay.Calls, Is.Zero);
            Assert.That(File.ReadAllBytes(temporary.Path), Is.EqualTo(before));
        });
    }

    [Test]
    public void Cbi48_C6_retry_or_abandonment_requires_explicit_reconciliation()
    {
        using var retryTemporary = new TemporaryJournal();
        var retry = Establish(retryTemporary.Path);
        var attempted = retry.BeginCycle().Snapshot;
        var ready = retry.ResolveInterrupted(ProviderTrustCadenceRecoveryDecision.Retry);

        using var abandonTemporary = new TemporaryJournal();
        var abandon = Establish(abandonTemporary.Path);
        abandon.BeginCycle();
        var abandoned = abandon.ResolveInterrupted(ProviderTrustCadenceRecoveryDecision.Abandon);

        Assert.Multiple(() =>
        {
            Assert.That(ready.Code, Is.EqualTo("durable-cadence-retry-ready"));
            Assert.That(ready.Snapshot.NextCycleIndex, Is.EqualTo(attempted.NextCycleIndex));
            Assert.That(ready.Snapshot.PreparedInstant, Is.EqualTo(attempted.PreparedInstant));
            Assert.That(ready.Snapshot.InterruptionCount, Is.EqualTo(1));
            Assert.That(ready.Snapshot.RetryCount, Is.EqualTo(1));
            Assert.That(abandoned.Code, Is.EqualTo("durable-cadence-abandoned"));
            Assert.That(abandoned.Snapshot.InterruptionCount, Is.EqualTo(1));
        });
    }

    [Test]
    public async Task Cbi48_C7_terminal_recovery_is_idempotent_and_effect_free()
    {
        using var temporary = new TemporaryJournal();
        var journal = Establish(temporary.Path);
        journal.BeginCycle();
        journal.CommitCycle("provider-trust-cycle-stopped");
        var before = File.ReadAllBytes(temporary.Path);
        var recovered = DurableProviderTrustCadenceJournal.Open(temporary.Path, RunId).Journal!;
        var cycle = new FakeCycle(_ => CycleResult("provider-trust-cycle-current"));
        var delay = new FakeDelay();

        var advanced = await ProviderTrustCadenceRecovery.AdvanceAsync(
            recovered, cycle, delay, CancellationToken.None);
        var reconciled = recovered.ResolveInterrupted(ProviderTrustCadenceRecoveryDecision.Retry);

        Assert.Multiple(() =>
        {
            Assert.That(advanced.Code, Is.EqualTo("durable-cadence-stopped"));
            Assert.That(reconciled.Code, Is.EqualTo("durable-cadence-stopped"));
            Assert.That(cycle.Calls, Is.Zero);
            Assert.That(delay.Calls, Is.Zero);
            Assert.That(File.ReadAllBytes(temporary.Path), Is.EqualTo(before));
        });
    }

    [Test]
    public void Cbi48_C8_reference_executes_the_shared_recovery_vectors()
    {
        using var fixture = JsonDocument.Parse(File.ReadAllText(Path.Combine(
            TestContext.CurrentContext.TestDirectory, "component-management", "fixtures",
            "cbi48-durable-cadence-vectors.json")));
        foreach (var vector in fixture.RootElement.GetProperty("vectors").EnumerateArray())
        {
            using var temporary = new TemporaryJournal();
            var journal = Establish(temporary.Path);
            foreach (var actionElement in vector.GetProperty("actions").EnumerateArray())
            {
                var action = actionElement.GetString()!;
                if (action.StartsWith("cycle:", StringComparison.Ordinal))
                {
                    journal.BeginCycle();
                    journal.CommitCycle(action[6..]);
                }
                else if (action == "gap")
                {
                    journal.CompleteGap(journal.Snapshot.PreparedInstant + Schedule.Interval);
                }
                else if (action == "crash") journal.BeginCycle();
                else if (action == "reopen")
                    journal = DurableProviderTrustCadenceJournal.Open(temporary.Path, RunId).Journal!;
                else if (action == "retry")
                    journal.ResolveInterrupted(ProviderTrustCadenceRecoveryDecision.Retry);
                else if (action == "abandon")
                    journal.ResolveInterrupted(ProviderTrustCadenceRecoveryDecision.Abandon);
                else Assert.Fail($"Unknown action {action}.");
            }

            var name = vector.GetProperty("name").GetString()!;
            var snapshot = journal.Snapshot;
            Assert.Multiple(() =>
            {
                Assert.That(snapshot.Code, Is.EqualTo(vector.GetProperty("expectedCode").GetString()), name);
                Assert.That(snapshot.Phase, Is.EqualTo(vector.GetProperty("expectedPhase").GetString()), name);
                Assert.That(snapshot.Cycles.Select(item => item.Code),
                    Is.EqualTo(vector.GetProperty("expectedCycleCodes").EnumerateArray()
                        .Select(item => item.GetString()).ToArray()), name);
                Assert.That(snapshot.Cycles.Select(item => item.Instant),
                    Is.EqualTo(vector.GetProperty("expectedCycleInstants").EnumerateArray()
                        .Select(item => item.GetDateTimeOffset()).ToArray()), name);
                Assert.That(snapshot.Gaps.Select(item => (int)item.TotalSeconds),
                    Is.EqualTo(vector.GetProperty("expectedGapsSeconds").EnumerateArray()
                        .Select(item => item.GetInt32()).ToArray()), name);
                Assert.That(snapshot.NextCycleIndex,
                    Is.EqualTo(vector.GetProperty("expectedNextCycle").GetInt32()), name);
                Assert.That(snapshot.InterruptionCount,
                    Is.EqualTo(vector.GetProperty("expectedInterruptions").GetInt32()), name);
                Assert.That(snapshot.RetryCount,
                    Is.EqualTo(vector.GetProperty("expectedRetries").GetInt32()), name);
            });
        }
    }
}

File: Minimal/tests/Brontide.Minimal.Host.Tests/ComponentProviderTrustCadenceRecoveryTests.fs
namespace Brontide.Minimal.Host.Tests

open System
open System.IO
open System.Text.Json
open System.Threading
open System.Threading.Tasks
open NUnit.Framework
open Brontide.Minimal.Host

type private Cbi48TemporaryJournal() =
    let root = Path.Combine(Path.GetTempPath(), $"brontide-cbi48-{Guid.NewGuid():N}")
    member _.Path = Path.Combine(root, "cadence.bin")
    interface IDisposable with
        member _.Dispose() =
            try if Directory.Exists root then Directory.Delete(root, true)
            with :? IOException | :? UnauthorizedAccessException -> ()

[<TestFixture>]
type ComponentProviderTrustCadenceRecoveryTests() =
    let runIdentity = ProviderTrustCadenceRunId.create "cadence-run.test.1"
    let schedule = ProviderServingTrustCadenceSchedule.create 2 (TimeSpan.FromSeconds 5.0)
    let start = DateTimeOffset(2026, 8, 5, 8, 0, 0, TimeSpan.Zero)
    let cycleResult code : ProviderServingTrustCycleResult =
        { Code = code
          Poll = None
          Sweep = None
          ServingCount = 0
          Rotation = None
          Availability = None }
    let establish path =
        DurableProviderTrustCadenceJournal.Establish(path, runIdentity, schedule, start).Journal.Value
    let multiple action = Assert.Multiple(Action action)

    [<Test>]
    member _.``CBI48 C1 a durable run is bounded and distinctly identified``() =
        use temporary = new Cbi48TemporaryJournal()
        let established =
            DurableProviderTrustCadenceJournal.Establish(temporary.Path, runIdentity, schedule, start)
        let duplicate =
            DurableProviderTrustCadenceJournal.Establish(temporary.Path, runIdentity, schedule, start)
        let mismatch = DurableProviderTrustCadenceJournal.Open(
            temporary.Path, ProviderTrustCadenceRunId.create "cadence-run.other")
        multiple (fun () ->
            Assert.That(established.Code, Is.EqualTo "durable-cadence-established")
            Assert.That(established.Journal.Value.Snapshot.RunIdentity, Is.EqualTo runIdentity)
            Assert.That(duplicate.Code, Is.EqualTo "durable-cadence-already-exists")
            Assert.That(mismatch.Code, Is.EqualTo "durable-cadence-run-mismatch")
            Assert.Throws<ArgumentException>(Action(fun () ->
                DurableProviderTrustCadenceJournal.Open(
                    temporary.Path, Unchecked.defaultof<ProviderTrustCadenceRunId>) |> ignore))
            |> ignore)

    [<Test>]
    member _.``CBI48 C2 every transition is atomic and integrity checked``() =
        use temporary = new Cbi48TemporaryJournal()
        let journal = establish temporary.Path
        Assert.That(journal.BeginCycle().Code, Is.EqualTo "durable-cadence-cycle-started")
        let bytes = File.ReadAllBytes temporary.Path
        bytes[bytes.Length - 1] <- bytes[bytes.Length - 1] ^^^ 0xffuy
        File.WriteAllBytes(temporary.Path, bytes)
        Assert.That(DurableProviderTrustCadenceJournal.Open(temporary.Path, runIdentity).Code,
            Is.EqualTo "durable-cadence-corrupt")

    [<Test>]
    member _.``CBI48 C3 in flight state precedes the effectful cycle``() = task {
        use temporary = new Cbi48TemporaryJournal()
        let journal = establish temporary.Path
        let mutable calls = 0
        let mutable observed = ""
        let cycle: ProviderServingTrustCycle = fun _ _ ->
            calls <- calls + 1
            observed <- DurableProviderTrustCadenceJournal.Open(temporary.Path, runIdentity).Code
            Task.FromResult(cycleResult "provider-trust-cycle-stopped")
        let delay: ProviderServingTrustCadenceDelay = fun now duration _ -> Task.FromResult(now + duration)
        let! _ = ProviderTrustCadenceRecovery.advance journal cycle delay CancellationToken.None
        multiple (fun () ->
            Assert.That(observed, Is.EqualTo "durable-cadence-indeterminate")
            Assert.That(calls, Is.EqualTo 1)
            Assert.That(journal.Snapshot.Code, Is.EqualTo "durable-cadence-stopped"))

        use failedWriteTemporary = new Cbi48TemporaryJournal()
        let failedWrite = establish failedWriteTemporary.Path
        File.Delete failedWriteTemporary.Path
        Directory.CreateDirectory failedWriteTemporary.Path |> ignore
        let mutable forbiddenCalls = 0
        let forbiddenCycle: ProviderServingTrustCycle = fun _ _ ->
            forbiddenCalls <- forbiddenCalls + 1
            Task.FromResult(cycleResult "provider-trust-cycle-current")
        let! refused =
            ProviderTrustCadenceRecovery.advance failedWrite forbiddenCycle delay CancellationToken.None
        multiple (fun () ->
            Assert.That(refused.Code, Is.EqualTo "durable-cadence-write-failed")
            Assert.That(forbiddenCalls, Is.Zero)) }

    [<Test>]
    member _.``CBI48 C4 completed work resumes from the next clean boundary``() = task {
        use temporary = new Cbi48TemporaryJournal()
        let first = establish temporary.Path
        let mutable firstCalls = 0
        let firstCycle: ProviderServingTrustCycle = fun _ _ ->
            firstCalls <- firstCalls + 1
            Task.FromResult(cycleResult "provider-trust-cycle-current")
        let noDelay: ProviderServingTrustCadenceDelay = fun now duration _ -> Task.FromResult(now + duration)
        let! _ = ProviderTrustCadenceRecovery.advance first firstCycle noDelay CancellationToken.None

        let waitingBytes = File.ReadAllBytes temporary.Path
        use canceled = new CancellationTokenSource()
        canceled.Cancel()
        let mutable canceledCycleCalls = 0
        let mutable canceledDelayCalls = 0
        let canceledCycle: ProviderServingTrustCycle = fun _ _ ->
            canceledCycleCalls <- canceledCycleCalls + 1
            Task.FromResult(cycleResult "provider-trust-cycle-current")
        let canceledDelay: ProviderServingTrustCadenceDelay = fun now duration _ ->
            canceledDelayCalls <- canceledDelayCalls + 1
            Task.FromResult(now + duration)
        let! canceledResult =
            ProviderTrustCadenceRecovery.advance first canceledCycle canceledDelay canceled.Token
        multiple (fun () ->
            Assert.That(canceledResult.Code, Is.EqualTo "durable-cadence-wait-canceled")
            Assert.That(canceledCycleCalls, Is.Zero)
            Assert.That(canceledDelayCalls, Is.Zero)
            Assert.That(File.ReadAllBytes temporary.Path, Is.EqualTo(box waitingBytes)))

        let recovered = DurableProviderTrustCadenceJournal.Open(temporary.Path, runIdentity).Journal.Value
        let mutable secondCalls = 0
        let mutable delayCalls = 0
        let secondCycle: ProviderServingTrustCycle = fun _ _ ->
            secondCalls <- secondCalls + 1
            Task.FromResult(cycleResult "provider-trust-cycle-current")
        let delay: ProviderServingTrustCadenceDelay = fun now duration _ ->
            delayCalls <- delayCalls + 1
            Task.FromResult(now + duration)
        let! _ = ProviderTrustCadenceRecovery.advance recovered secondCycle delay CancellationToken.None
        multiple (fun () ->
            Assert.That(firstCalls, Is.EqualTo 1)
            Assert.That(secondCalls, Is.EqualTo 1)
            Assert.That(delayCalls, Is.EqualTo 1)
            Assert.That(recovered.Snapshot.Code, Is.EqualTo "durable-cadence-complete")
            Assert.That(recovered.Snapshot.Cycles |> List.map _.Index, Is.EqualTo(box [ 0; 1 ]))) }

    [<Test>]
    member _.``CBI48 C5 an interrupted effect is indeterminate and inert``() = task {
        use temporary = new Cbi48TemporaryJournal()
        let journal = establish temporary.Path
        journal.BeginCycle() |> ignore
        let before = File.ReadAllBytes temporary.Path
        let recovered = DurableProviderTrustCadenceJournal.Open(temporary.Path, runIdentity)
        let mutable cycleCalls = 0
        let mutable delayCalls = 0
        let cycle: ProviderServingTrustCycle = fun _ _ ->
            cycleCalls <- cycleCalls + 1
            Task.FromResult(cycleResult "provider-trust-cycle-current")
        let delay: ProviderServingTrustCadenceDelay = fun now duration _ ->
            delayCalls <- delayCalls + 1
            Task.FromResult(now + duration)
        let! result = ProviderTrustCadenceRecovery.advance recovered.Journal.Value cycle delay CancellationToken.None
        multiple (fun () ->
            Assert.That(recovered.Code, Is.EqualTo "durable-cadence-indeterminate")
            Assert.That(result.Code, Is.EqualTo "durable-cadence-indeterminate")
            Assert.That(cycleCalls, Is.Zero)
            Assert.That(delayCalls, Is.Zero)
            Assert.That(File.ReadAllBytes temporary.Path, Is.EqualTo(box before))) }

    [<Test>]
    member _.``CBI48 C6 retry or abandonment requires explicit reconciliation``() =
        use retryTemporary = new Cbi48TemporaryJournal()
        let retry = establish retryTemporary.Path
        let attempted = retry.BeginCycle().Snapshot
        let ready = retry.ResolveInterrupted ProviderTrustCadenceRecoveryDecision.Retry
        use abandonTemporary = new Cbi48TemporaryJournal()
        let abandon = establish abandonTemporary.Path
        abandon.BeginCycle() |> ignore
        let abandoned = abandon.ResolveInterrupted ProviderTrustCadenceRecoveryDecision.Abandon
        multiple (fun () ->
            Assert.That(ready.Code, Is.EqualTo "durable-cadence-retry-ready")
            Assert.That(ready.Snapshot.NextCycleIndex, Is.EqualTo attempted.NextCycleIndex)
            Assert.That(ready.Snapshot.PreparedInstant, Is.EqualTo attempted.PreparedInstant)
            Assert.That(ready.Snapshot.InterruptionCount, Is.EqualTo 1)
            Assert.That(ready.Snapshot.RetryCount, Is.EqualTo 1)
            Assert.That(abandoned.Code, Is.EqualTo "durable-cadence-abandoned")
            Assert.That(abandoned.Snapshot.InterruptionCount, Is.EqualTo 1))

    [<Test>]
    member _.``CBI48 C7 terminal recovery is idempotent and effect free``() = task {
        use temporary = new Cbi48TemporaryJournal()
        let journal = establish temporary.Path
        journal.BeginCycle() |> ignore
        journal.CommitCycle "provider-trust-cycle-stopped" |> ignore
        let before = File.ReadAllBytes temporary.Path
        let recovered = DurableProviderTrustCadenceJournal.Open(temporary.Path, runIdentity).Journal.Value
        let mutable cycleCalls = 0
        let mutable delayCalls = 0
        let cycle: ProviderServingTrustCycle = fun _ _ ->
            cycleCalls <- cycleCalls + 1
            Task.FromResult(cycleResult "provider-trust-cycle-current")
        let delay: ProviderServingTrustCadenceDelay = fun now duration _ ->
            delayCalls <- delayCalls + 1
            Task.FromResult(now + duration)
        let! advanced = ProviderTrustCadenceRecovery.advance recovered cycle delay CancellationToken.None
        let reconciled = recovered.ResolveInterrupted ProviderTrustCadenceRecoveryDecision.Retry
        multiple (fun () ->
            Assert.That(advanced.Code, Is.EqualTo "durable-cadence-stopped")
            Assert.That(reconciled.Code, Is.EqualTo "durable-cadence-stopped")
            Assert.That(cycleCalls, Is.Zero)
            Assert.That(delayCalls, Is.Zero)
            Assert.That(File.ReadAllBytes temporary.Path, Is.EqualTo(box before))) }

    [<Test>]
    member _.``CBI48 C8 minimal executes the shared recovery vectors``() =
        use fixture = JsonDocument.Parse(File.ReadAllText(Path.Combine(
            TestContext.CurrentContext.TestDirectory, "component-management", "fixtures",
            "cbi48-durable-cadence-vectors.json")))
        let textValue (value: JsonElement) = value.GetString() |> Option.ofObj |> Option.defaultValue ""
        for vector in fixture.RootElement.GetProperty("vectors").EnumerateArray() do
            use temporary = new Cbi48TemporaryJournal()
            let mutable journal = establish temporary.Path
            for actionElement in vector.GetProperty("actions").EnumerateArray() do
                let action = textValue actionElement
                if action.StartsWith("cycle:", StringComparison.Ordinal) then
                    journal.BeginCycle() |> ignore
                    journal.CommitCycle(action[6..]) |> ignore
                elif action = "gap" then
                    journal.CompleteGap(journal.Snapshot.PreparedInstant + schedule.Interval) |> ignore
                elif action = "crash" then journal.BeginCycle() |> ignore
                elif action = "reopen" then
                    journal <- DurableProviderTrustCadenceJournal.Open(temporary.Path, runIdentity).Journal.Value
                elif action = "retry" then
                    journal.ResolveInterrupted ProviderTrustCadenceRecoveryDecision.Retry |> ignore
                elif action = "abandon" then
                    journal.ResolveInterrupted ProviderTrustCadenceRecoveryDecision.Abandon |> ignore
                else Assert.Fail $"Unknown action {action}."
            let name = vector.GetProperty("name") |> textValue
            let snapshot = journal.Snapshot
            multiple (fun () ->
                Assert.That(snapshot.Code, Is.EqualTo(vector.GetProperty("expectedCode") |> textValue), name)
                Assert.That(snapshot.Phase, Is.EqualTo(vector.GetProperty("expectedPhase") |> textValue), name)
                Assert.That(snapshot.Cycles |> List.map _.Code, Is.EqualTo(box (
                    vector.GetProperty("expectedCycleCodes").EnumerateArray()
                    |> Seq.map textValue |> Seq.toList)), name)
                Assert.That(snapshot.Cycles |> List.map _.Instant, Is.EqualTo(box (
                    vector.GetProperty("expectedCycleInstants").EnumerateArray()
                    |> Seq.map _.GetDateTimeOffset() |> Seq.toList)), name)
                Assert.That(snapshot.Gaps |> List.map (fun gap -> int gap.TotalSeconds), Is.EqualTo(box (
                    vector.GetProperty("expectedGapsSeconds").EnumerateArray()
                    |> Seq.map _.GetInt32() |> Seq.toList)), name)
                Assert.That(snapshot.NextCycleIndex,
                    Is.EqualTo(vector.GetProperty("expectedNextCycle").GetInt32()), name)
                Assert.That(snapshot.InterruptionCount,
                    Is.EqualTo(vector.GetProperty("expectedInterruptions").GetInt32()), name)
                Assert.That(snapshot.RetryCount,
                    Is.EqualTo(vector.GetProperty("expectedRetries").GetInt32()), name))
